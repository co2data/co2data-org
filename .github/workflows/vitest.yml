name: Unit tests
on: push

concurrency: ${{ github.workflow }}-${{ github.ref }}

jobs:
  unit-test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: pnpm/action-setup@v2
        with:
          version: 8
      - uses: actions/setup-node@v3
        with:
          node-version: 20
          cache: 'pnpm'

      - run: pnpm install

      - name: Run unit tests
        run: pnpm test
