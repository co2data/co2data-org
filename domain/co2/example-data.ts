import { Co2Average } from '.'

export function makeCo2Average() {
  return {
    id: 'id',
    title: 'Pork',
    slug: 'test',
    unit: 'kilogram',
    avgPerYear: 500000,
    avgPerUnit: 30000,
    singleConsumptionFrom: 0,
    singleConsumptionAverage: 4,
    singleConsumptionTo: 100,
    timesPerYearFrom: 0,
    timesPerYearAverage: 5,
    timesPerYearTo: 20,
  } satisfies Co2Average
}
