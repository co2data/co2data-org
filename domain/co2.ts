import { Co2RepositoryLive } from '@/adapter/co2-repository'
import { DbLive } from '@/infra/db'
import { Layer } from 'effect'
export { Co2Repository } from '@/adapter/co2-repository'
export type Unit =
  | 'gram'
  | 'hour'
  | 'kilogram'
  | 'kilometer'
  | 'liter'
  | 'meter'
  | 'minute'

export type Co2Average = {
  id: string
  title: string
  slug: string
  unit: Unit
  avgPerYear: number
  avgPerUnit: number
  singleConsumptionFrom: number
  singleConsumptionTo: number
  singleConsumptionAverage: number
  timesPerYearFrom: number
  timesPerYearTo: number
  timesPerYearAverage: number
}

export const repositoryLive = Co2RepositoryLive.pipe(Layer.use(DbLive))
