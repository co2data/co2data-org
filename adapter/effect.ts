import { Effect, Layer, flow } from 'effect'
import { Co2Repository } from './co2-repository'
import { SourceRepository, SourceRepositoryLive } from './source-repository'
import { NodeSdkLive } from '@/infra/tracing/NodeSdkLive'
import { sourceRepoLive } from '@/domain/source'
import { co2RepoLive } from '@/domain/co2'

export function run<P, Q, A>(
  effect: (
    arg1: P,
    arg2: Q
  ) => Effect.Effect<Co2Repository | SourceRepository, never, A>
) {
  return flow(
    effect,
    Effect.provide(Layer.merge(co2RepoLive, sourceRepoLive)),
    Effect.provide(NodeSdkLive),
    Effect.runPromise
  )
}
