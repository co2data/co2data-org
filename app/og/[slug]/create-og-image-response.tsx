import OgImageFrame from '@/components/og-image-frame'
import { Co2Average, Co2Repository } from '@/domain/co2'
import { DbError } from '@/infra/db'
import { BaseError } from '@/lib/types'
import convert from 'convert'
import { Data, Effect } from 'effect'
import { ImageResponse } from 'next/server'
import { Fragment } from 'react'

export function createOgImageResponse(
  request: Request,
  { params }: { params: { slug: string | undefined } }
) {
  return parseSlug(params.slug).pipe(
    Effect.flatMap(getCo2Average),
    Effect.flatMap(renderImageAsResponse),
    Effect.catchAll(handleErrors)
  ) satisfies Effect.Effect<any, never, ImageResponse>
}

export function renderImageAsResponse(co2Avg: Co2Average) {
  const { formattedInteger, formattedAvgRest } = renderNumberParts(co2Avg)
  const renderedImage = renderImage(co2Avg, formattedInteger, formattedAvgRest)
  return wrapInResponse(renderedImage).pipe(Effect.withSpan('/render image'))
}
function renderNumberParts(co2Avg: Co2Average) {
  const formattedParts = formatter.formatToParts(
    convert(co2Avg.avgPerUnit, 'grams').to('kg')
  )

  const formattedInteger = formattedParts.find(
    (part) => part.type === 'integer'
  )?.value
  const formattedAvgRest = formattedParts.map(({ type, value }) => {
    switch (type) {
      case 'integer':
      case 'unit':
      case 'literal':
        return <Fragment key={type} />
      default:
        return (
          <div key={type} tw="flex text-8xl leading-[0.8] mt-1">
            {value}
          </div>
        )
    }
  })
  return { formattedInteger, formattedAvgRest }
}

export function renderImage(
  co2Avg: Co2Average,
  formattedInteger: string | undefined,
  formattedAvgRest: JSX.Element[]
) {
  return (
    <OgImageFrame>
      <div tw="flex justify-between items-end mt-8">
        <div tw="flex flex-col shrink flex-1 -mb-1">
          <div tw="flex flex-col text-4xl">Emissions of 1 {co2Avg.unit}</div>
          <div tw="flex flex-col text-8xl leading-[0.9] -ml-1">
            {co2Avg.title}
          </div>
        </div>
        <div tw="flex shrink-0 text-[42] leading-[0.75] ml-6">
          {formattedInteger}
        </div>
        <div tw="flex flex-col text-4xl justify-end">
          <div tw="flex flex-col ml-2 self-end">kg CO2e</div>
          <div tw="flex">{formattedAvgRest}</div>
        </div>
      </div>
    </OgImageFrame>
  )
}

export function wrapInResponse(renderedImage: JSX.Element) {
  return Effect.try({
    try: () =>
      new ImageResponse(renderedImage, {
        width: 1200,
        height: 630,
      }),
    catch: (cause) => new ImageRenderError({ cause }),
  })
}

const formatter = new Intl.NumberFormat('en', {
  style: 'unit',
  unit: 'kilogram',
  notation: 'compact',
})

function parseSlug(slug: string | undefined) {
  return Effect.fromNullable(slug).pipe(
    Effect.mapError(() => new NoParamError())
  )
}

function getCo2Average(slug: string) {
  return Co2Repository.pipe(
    Effect.flatMap((repo) => repo.getCo2AverageBySlug(slug)),
    Effect.flatMap(Effect.orElseFail(notFound(slug)))
  )
}
function notFound(slug: string | undefined) {
  return () =>
    new NotFoundError({
      cause: `Entry with slug ${slug} not found.`,
    })
}

class NoParamError extends Data.TaggedError('NoParam')<{}> {
  readonly cause = 'No parameter defined.'
}
class NotFoundError extends Data.TaggedError('NotFound')<BaseError> {}

class ImageRenderError extends Data.TaggedError('ImageRender')<BaseError> {}

function handleErrors(
  error: NoParamError | DbError | NotFoundError | ImageRenderError
) {
  return Effect.succeed(
    new Response(`Failed to generate the image. ${error.cause}`, {
      status: 500,
    })
  )
}
