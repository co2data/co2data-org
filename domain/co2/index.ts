import { Option } from 'effect'

export { Co2Repository } from '@/domain/co2/repository'
export type Unit =
  | 'gram'
  | 'hour'
  | 'kilogram'
  | 'kilometer'
  | 'liter'
  | 'meter'
  | 'minute'

export type Co2Average = {
  id: string
  title: string
  description: Option.Option<string>
  slug: string
  unit: Unit
  avgPerYear: number
  avgPerUnit: number
  singleConsumptionFrom: number
  singleConsumptionTo: number
  singleConsumptionAverage: number
  timesPerYearFrom: number
  timesPerYearTo: number
  timesPerYearAverage: number
}
