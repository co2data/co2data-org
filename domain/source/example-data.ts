import { Option } from 'effect'
import type { Source } from '.'

export function makeSource(override: Partial<Source> = {}) {
  return {
    id: '1',
    name: 'Sustainability Report 2018',
    description: 'No description',
    gCo2e: 2000,
    per: 1,
    year: 2018,
    userId: 'user1',
    links: Option.some([
      {
        id: 'id1',
        name: 'Excel sheet',
        mediaType: 'xls',
        url: 'https://some-server.domain/route.xls',
      },
    ]),
    region: 'World',
    ...override,
  } satisfies Source
}
