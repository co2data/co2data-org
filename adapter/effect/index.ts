import { NodeSdkLive } from '@/adapter/tracing/NodeSdkLive'
import { Effect, flow } from 'effect'
import 'server-only'
import { Co2Repository } from '../../domain/co2/repository'
import { SourceRepository } from '../../domain/source/repository'
import { mainLive } from './main'

export function run<P, Q, A>(
  effect: (
    arg1: P,
    arg2: Q
  ) => Effect.Effect<Co2Repository | SourceRepository, never, A>
) {
  return flow(
    effect,
    Effect.provide(mainLive),
    Effect.provide(NodeSdkLive),
    Effect.runPromise
  )
}
