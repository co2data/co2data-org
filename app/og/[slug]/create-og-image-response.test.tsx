import { render, screen } from '@testing-library/react'
import { describe, expect, it, test } from 'vitest'
import { renderImage } from './create-og-image-response'
import { makeCo2Average } from '@/domain/co2/example-data'
import { Effect } from 'effect'

describe('og image test', () => {
  it('render image', () => {
    const actual = renderImage(makeCo2Average(), '3', [
      <div key="hi">4</div>,
      <div key="ho">500</div>,
    ])

    expect(actual).matchSnapshot('image image')
  })
})
