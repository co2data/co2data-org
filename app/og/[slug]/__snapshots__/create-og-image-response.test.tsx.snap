// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`og image test > render image > image image 1`] = `
<OgImageFrame>
  <div
    tw="flex justify-between items-end mt-8"
  >
    <div
      tw="flex flex-col shrink flex-1 -mb-1"
    >
      <div
        tw="flex flex-col text-4xl"
      >
        Emissions of 1 
        kilogram
      </div>
      <div
        tw="flex flex-col text-8xl leading-[0.9] -ml-1"
      >
        Pork
      </div>
    </div>
    <div
      tw="flex shrink-0 text-[42] leading-[0.75] ml-6"
    >
      3
    </div>
    <div
      tw="flex flex-col text-4xl justify-end"
    >
      <div
        tw="flex flex-col ml-2 self-end"
      >
        kg CO2e
      </div>
      <div
        tw="flex"
      >
        <div>
          4
        </div>
        <div>
          500
        </div>
      </div>
    </div>
  </div>
</OgImageFrame>
`;
